Source/WebCore/editing/TextIterator.cpp

```diff
@@ -916,23 +916,16 @@ static bool shouldEmitNewlineBeforeNode(Node& node)
 
 static bool shouldEmitExtraNewlineForNode(Node& node)
 {
-    // When there is a significant collapsed bottom margin, emit an extra
-    // newline for a more realistic result. We end up getting the right
-    // result even without margin collapsing. For example: <div><p>text</p></div>
-    // will work right even if both the <div> and the <p> have bottom margins.
+    // https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute
+    // Append two required linebreaks after a <p> element.
 
     CheckedPtr renderBox = dynamicDowncast<RenderBox>(node.renderer());
     if (!renderBox || !renderBox->height())
         return false;
 
-    // NOTE: We only do this for a select set of nodes, and WinIE appears not to do this at all.
-    RefPtr element = dynamicDowncast<HTMLElement>(node);
-    if (!element || (!hasHeaderTag(*element) && !is<HTMLParagraphElement>(*element)))
-        return false;
-
-    auto bottomMargin = renderBox->collapsedMarginAfter();
-    auto fontSize = renderBox->style().fontDescription().computedSize();
-    return bottomMargin * 2 >= fontSize;
+    if (RefPtr element = dynamicDowncast<HTMLElement>(node))
```

Comment:
Or it could be:

```
return node.hasTagName(pTag);

```

OR

```
RefPtr element = dynamicDowncast<HTMLElement>(node);
if (!element)
    return false;
return element->hasTagName(pTag);

```

Reply:
@rniwa - C++ change (only one file and function touched):

<img width="1197" alt="image" src="https://github.com/WebKit/WebKit/assets/52317531/03130e12-3b49-4de7-b47e-a8471e6fd1bd">